grabbit/core.py

```diff
@@ -155,43 +153,81 @@ def __init__(self, path, config=None, dynamic_getters=False,
+            if k in config:
+                config[k] = v
+
         for e in config['entities']:
             self.add_entity(**e)
 
+        if 'index' in config:
+            self.inclusion_regex = config['index']
+            if self.inclusion_regex.get('include') and \
+                self.inclusion_regex.get('exclude'):
+                    raise ValueError("You can only define either include or "
+                                     "exclude regex, not both.")
         self.index()
 
+    def _check_inclusions(self, f):
+        ''' Check if file or directory against regexes in config to determine if
+            it should be included in the index '''
+        filename = f if isinstance(f, str) else f.filename
+
+        # If file matches any include regex, then true
+        include_regex = self.inclusion_regex.get('include', [])
+        if include_regex:
+            for regex in include_regex:
+                if re.match(regex, filename):
+                    break
+            else:
+                return False
+
+        # If file matches any excldue regex, then false
+        for regex in self.inclusion_regex.get('exclude', []):
+            if re.match(regex ,filename):
+                return False
+
+        return True
+
+    def _validate_dir(self, d):
+        ''' Extend this in subclasses to provide additional directory validation.
+        Will be called the first time a directory is read in; if False is
+        returned, the directory will be ignored and dropped from the layout. '''
+
+        return True
+
     def _validate_file(self, f):
-        ''' Override this in subclasses to provide additional file validation.
+        ''' Extend this in subclasses to provide additional file validation.
         Will be called the first time each file is read in; if False is
         returned, the file will be ignored and dropped from the layout. '''
+
         return True
 
     def index(self):
-
         # Reset indexes
         self.files = {}
         for ent in self.entities.values():
             ent.files = {}
 
         # Loop over all files
         for root, directories, filenames in os.walk(self.root, topdown=True):
-
-            # Exclude directories from further search if they match exclude regex
-            if self.exclude_dir is not None:
-                directories[:] = [d for d in directories if not re.match(self.exclude_dir, d)]
-
+            # Only loop further over directories that pass regex (on full path)
+            full_dirs = [os.path.join(root, d) for d in directories]
+            full_dirs = filter(self._check_inclusions, full_dirs)
+            directories[:] = [os.path.split(d)[1] for d in \
+                              filter(self._validate_dir, full_dirs)]
             for f in filenames:
                 f = File(join(root, f))
-                if not self._validate_file(f):
+                if not (self._check_inclusions(f) or self._validate_file(f)):
```

Comment:
Should be `and` instead of `or`, no? (Or alternatively, drop the parentheses.) We want to skip file if either condition fails.